feat(GUI): separate svg path and content attributes

Shou opened the pull request:
We separate the SVG component path and content into attributes
`paths` and `contents` which take lists of strings that are
tried until one succeeds. `contents` takes precedence over `paths`,
i.e. it is tried first.

Change-Type: patch
Changelog-Entry: Separate SVG component's path and content attributes.

lurch commented on tests/gui/components/svg-icon.spec.js:
```diff
@@ -48,7 +48,7 @@ describe('Browser: SVGIcon', function () {
       iconContents = iconContents.join('\n')
 
-      const element = $compile(`<svg-icon path="'${icon}'">Resin.io</svg-icon>`)($rootScope)
+      const element = $compile(`<svg-icon paths="['${icon}']">Resin.io</svg-icon>`)($rootScope)
```
Hmmm, so it looks like `paths` and `contents` both now actually accept arrays of values, rather than just a single value? And it then searches through each of these arrays, and displays the first value in the array that's valid? Sounds like lots more unit tests would be useful to validate this new behaviour... ;-)

(and this behaviour should also be mentioned in the jsdoc, especially the fact that 'contents' (if present) is checked before `paths` is checked)

Shou replied:
I've added some tests for this now

lurch commented on lib/gui/pages/main/templates/main.tpl.html:
```diff
@@ -3,7 +3,7 @@
 
       <div class="center-block">
-        <svg-icon path="main.selection.getImageLogo() || '../assets/image.svg'"></svg-icon>
+        <svg-icon contents="main.selection.getImageLogo()" path="'../assets/image.svg'"></svg-icon>
```
I'm guessing that `path=` here should be `paths=` ? And should these also be using array-notation like the tests are using?

Shou replied:
That's true, I've fixed the `path` into `paths` now

lurch commented on lib/gui/pages/main/templates/main.tpl.html:
```diff
@@ -3,7 +3,7 @@
 
       <div class="center-block">
-        <svg-icon path="main.selection.getImageLogo() || '../assets/image.svg'"></svg-icon>
+        <svg-icon contents="main.selection.getImageLogo()" path="'../assets/image.svg'"></svg-icon>
```
Hmmm, as well as `paths` and `contents` I wonder if it'd be worth also allowing `content` and `path` for the (common) situation where you only want to pass in a single path, not multiple paths? \*shrug\*
(obviously you'd need to explicitly document which order each of these properties are tried in)

jhermsmeier commented:
@Shou what's the state of this?

Shou commented:
I've just rebased and fixed this, [here are the relevant changes][1]. Let me know what you think.

[1]: https://github.com/resin-io/etcher/compare/6b0ed64819cd1a178d18e9fd0605e98f69eb8b8a...svg-contents-attribute

jhermsmeier commented:
@Shou looks like the CIs aren't happy

jhermsmeier commented:
Ping @Shou

Shou commented:
Rebased on master and fixed linter errors

jhermsmeier commented on lib/gui/app/components/svg-icon.js:
```diff
@@ -59,31 +83,52 @@ class SVGIcon extends react.Component {
+        // relative to *this directory*.
+        // TODO: There might be a way to compute the path
+        // relatively to the `index.html`.
+        const imagePath = path.join(baseDirectory, 'assets', relativePath)
```
Regarding the TODO here, I think using `path.relative(from,to)` could work here

Shou replied:
Let's fix that in a separate PR, as this is just a moved block, the TODO isn't new to this PR.

jhermsmeier replied:
Oh, right – yeah, let's do that then

Shou replied:
Sweet, is this good to merge then?

jhermsmeier commented:
What about @lurch's comment?

> Looks like this PR also adds the "allow multiple paths to be specified so that we can give fallback values" - IMHO that should be mentioned in the commit message too.

Shou commented:
@jhermsmeier updated the commit message and the OP as well

lurch commented on lib/gui/app/components/svg-icon.js:
```diff
@@ -59,31 +83,52 @@ class SVGIcon extends react.Component {
+          return false
+        }
+
+        const parsed = _.attempt(tryParseSVGContents, contents)
```
Does this _need_ to use `_.attempt`, given that the call on line 89 doesn't?

lurch commented on lib/gui/app/components/svg-icon.js:
```diff
@@ -59,31 +83,52 @@ class SVGIcon extends react.Component {
-    const svgXml = svg && _.isNil(parserError) ? svg.outerHTML : ''
-    const svgData = `data:image/svg+xml,${encodeURIComponent(svgXml)}`
-
     return react.createElement('img', {
```
What's the behaviour if neither `contents` or `paths` resolves to an actual SVG, i.e. `svgData` is still set to `''`?

lurch commented on lib/gui/app/components/svg-icon.js:
```diff
@@ -108,9 +153,14 @@ class SVGIcon extends react.Component {
+   * @summary List of embedded SVG contents to be tried in succession if any fails
    */
-  path: propTypes.string.isRequired,
+  contents: propTypes.array,
```
Minor nit, but given that the code attempts to parse `contents` before parsing `paths`, maybe `contents` should be declared above `paths`?

lurch commented on lib/gui/app/pages/main/templates/main.tpl.html:
```diff
@@ -3,7 +3,7 @@
 
       <div class="center-block">
-        <svg-icon path="main.selection.getImageLogo() || '../../assets/image.svg'"></svg-icon>
+        <svg-icon contents="main.selection.getImageLogo()" paths="[ '../../assets/image.svg' ]"></svg-icon>
```
Isn't `contents` supposed to be an array?! See e.g. https://github.com/resin-io/etcher/pull/1677/files#diff-0bee66f458b50bfd24c37224e5891e78R105

jhermsmeier commented on lib/gui/app/components/svg-icon.js:
```diff
@@ -37,6 +38,29 @@ const DEFAULT_SIZE = '40px'
+const tryParseSVGContents = (contents) => {
+  const doc = domParser.parseFromString(contents, 'image/svg+xml')
+  const parserError = doc.querySelector('parsererror')
+  const svg = doc.querySelector('svg')
```
@lurch it's counter-intuitive, but a parser error does not mean that no document is constructed; in the case of SVG, there will at least be an "error document", which is an SVG containing the error message

lurch replied:
Okay. Perhaps I should have left this comment on this line instead ;)  https://github.com/resin-io/etcher/pull/1677/files#diff-9d4c83308fdfbb2b3567f9a9ecff3136R57
(but as I said, only a minor nit)
